uaclient/tests/test_contract.py

```diff
@@ -260,6 +262,80 @@ def fake_contract_client(cfg):
 
         assert "Broken ent1 route" == str(exc.value)
 
+    @pytest.mark.parametrize(
+        "error_instance, ux_error_msg",
+        (
+            (
+                exceptions.UserFacingError(
+                    "Ubuntu Advantage server provided no aptKey directive for"
+                    " esm-infra"
+                ),
+                MESSAGE_ATTACH_FAILURE_DEFAULT_SERVICES,
+            ),
+            (
+                RuntimeError("some APT error"),
+                MESSAGE_UNEXPECTED_ERROR_DURING_OP_TMPL.format(
+                    operation="attach"
+                ),
+            ),
+        ),
+    )
+    @mock.patch(M_PATH + "process_entitlement_delta")
+    @mock.patch("uaclient.util.get_machine_id", return_value="mid")
+    @mock.patch(M_PATH + "UAContractClient")
+    def test_user_facing_error_due_to_unexpected_process_entitlement_delta(
+        self,
+        client,
+        get_machine_id,
+        process_entitlement_delta,
+        error_instance,
+        ux_error_msg,
+    ):
+        """Unexpected errors from process_entitlement_delta are raised.
+
+        Remaining entitlements are processed regardless of error and error is
+        raised at the end.
+        """
+        # Fail first and succeed second call to process_entitlement_delta
+        process_entitlement_delta.side_effect = (error_instance, None)
+
+        # resourceEntitlements specifically ordered reverse alphabetically
+        # to ensure proper sorting for process_contract_delta calls below
+        machine_token = {
+            "machineToken": "mToken",
+            "machineTokenInfo": {
+                "contractInfo": {
+                    "id": "cid",
+                    "resourceEntitlements": [
+                        {"entitled": False, "type": "ent2"},
+                        {"entitled": True, "type": "ent1"},
+                    ],
+                }
+            },
+        }
+
+        def fake_contract_client(cfg):
+            client = FakeContractClient(cfg)
+            client._responses = {
+                self.refresh_route: machine_token,
+                self.access_route_ent1: {
+                    "entitlement": {
+                        "entitled": True,
+                        "type": "ent1",
+                        "new": "newval",
+                    }
+                },
+            }
+            return client
+
+        client.side_effect = fake_contract_client
```

Comment:
Is it possible to simplify this by using `return_value` instead of `side_effect`?

Reply:
I could mock return values for the two UAContractClient.request_* methods that are called for self.refresh_route and self.access_route_ent1  but then I'd also have to mock the cfg.write_cache calls when seed the for loop that iterates over cfg.entitlements.items().  So it adds complexity there because we also need to properly setup the cfg.entitlements values too.

Reply:
Apologies, my question was not really specific enough at all, I was really asking if this alternative, slightly simpler implementation would work:

```py
        fake_contract_client = FakeContractClient(cfg)
        fake_contract_client._responses = {
            self.refresh_route: machine_token,
            self.access_route_ent1: {
                "entitlement": {
                    "entitled": True,
                    "type": "ent1",
                    "new": "newval",
                }
            },
        }

        client.return_value = fake_contract_client
```